helix-core/src/main/java/org/apache/helix/controller/stages/BestPossibleStateCalcStage.java

```diff
@@ -131,82 +135,102 @@ public void process(ClusterEvent event) throws Exception {
     });
   }
 
+  private String selectSwapInState(StateModelDefinition stateModelDef, Map<String, String> stateMap,
+      String swapOutInstance) {
+    // If the swap-in node is live, select state with the following logic:
+    // 1. If the swap-out instance's replica is in the stateMap:
+    // - if the swap-out instance's replica is a topState, select the swap-in instance's replica to the topState.
+    //   if another is allowed to be added, otherwise select the swap-in instance's replica to a secondTopState.
+    // - if the swap-out instance's replica is not a topState or ERROR, select the swap-in instance's replica to the same state.
+    // - if the swap-out instance's replica is ERROR, select the swap-in instance's replica to the initialState.
+    // 2. If the swap-out instance's replica is not in the stateMap, select the swap-in instance's replica to the initialState.
+    // This happens when the swap-out node is offline.
+    if (stateMap.containsKey(swapOutInstance)) {
+      if (stateMap.get(swapOutInstance).equals(stateModelDef.getTopState()) || stateMap.get(
+          swapOutInstance).equals(HelixDefinedState.ERROR.name())) {
+        // If the swap-out instance's replica is a topState, select the swap-in instance's replica
+        // to be the topState if the StateModel allows another to be added. If not, select the swap-in
+        // to be the secondTopState.
+        String topStateCount = stateModelDef.getNumInstancesPerState(stateModelDef.getTopState());
+        if (topStateCount.equals(StateModelDefinition.STATE_REPLICA_COUNT_ALL_CANDIDATE_NODES)
+            || topStateCount.equals(StateModelDefinition.STATE_REPLICA_COUNT_ALL_REPLICAS)) {
+          // If the StateModel allows for another replica with the topState to be added,
+          // select the swap-in instance's replica to the topState.
+          return stateModelDef.getTopState();
+        } else {
+          // If StateModel does not allow another topState replica to be
+          // added, select the swap-in instance's replica to be the secondTopState.
+          return stateModelDef.getSecondTopStates().iterator().next();
+        }
+      } else {
+        // If the swap-out instance's replica is not a topState or ERROR, select the swap-in instance's replica
+        // to be the same state
+        return stateMap.get(swapOutInstance);
+      }
+    } else {
+      // If the swap-out instance's replica is not in the stateMap, select the swap-in instance's replica
+      // to be the initialState. This happens when the swap-out node is offline.
+      return stateModelDef.getInitialState();
```

Comment:
This is confusing to me. If the instance is not state map, it suppose not be in preference list. Also, the original swap out one is already dropped. Then we should not even to put a state to that instance, I think?

Reply:
It is possible for the instance to be in the preference list but for it not to be in the stateMap. This is when the swap out instance is offline. It will disappear from idealState. The assignment is still to the offline node.

Reply:
I changed it to only bootstrap replicas on swap in node based on the corresponding instances in the state map. This means that offline instances will not have its assigned partitions on the swap in node because we don't know what the right state would be to copy over.